02-peer-protocol.md

```diff
@@ -390,6 +400,377 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+     * [`shutdown_len`:`shutdown_scriptpubkey`]
+
+Rationale and Requirements are the same as listed above, for [`open_channel`](#the-open_channel-message), with the following addendums:
+
+#### Requirements:
+
+If nodes have negotiated `option_dual_fund`:
+  - the opening node:
+    - MUST not send `open_channel`
+
+#### Rationale
+
+`feerate_per_kw_funding`: fee rate that the opening node will pay
+for the funding transaction.
+
+
+### The `accept_channel2` Message
+
+This message contains information about a node and indicates its
+acceptance of the new channel.
+
+1. type: 33 (`accept_channel`)
+2. data:
+  * [`32`:`temporary_channel_id`]
+  * [`8`:`dust_limit_satoshis`]
+  * [`8`:`max_htlc_value_in_flight_msat`]
+  * [`8`:`channel_reserve_satoshis`]
+  * [`8`:`htlc_minimum_msat`]
+  * [`4`:`minimum_depth`]
+  * [`2`:`to_self_delay`]
+  * [`2`:`max_accepted_htlcs`]
+  * [`33`:`funding_pubkey`]
+  * [`33`:`revocation_basepoint`]
+  * [`33`:`payment_basepoint`]
+  * [`33`:`delayed_payment_basepoint`]
+  * [`33`:`htlc_basepoint`]
+  * [`33`:`first_per_commitment_point`]
+  * [`2`:`opening_tlv_len`]
+  * [`opening_tlv_len`:`opening_tlv`]
+
+Rationale and Requirements are the same as listed above,
+for [`accept_channel`](#the-accept_channel-message)
+
+### The `funding_compose` Message
+
+This message exchanges the transaction input and output
+information necessary to compose the funding transaction.
+
+1. type: 58 (`funding_compose`)
+2. data:
+  * [`32`:`temporary_channel_id`]
+  * [`2`:`num_inputs`]
+  * [`num_inputs`:`input_info`]
+  * [`2`:`num_outputs`]
+  * [`num_outputs`:`output_info`]
+
+1. subtype `input_info`
+2. data:
+  * [`8`:`satoshis`]
+  * [`32`:`prevtxid`]
+  * [`4`:`prevtxoutnum`]
+  * [`2`:`max_witness_len`]
+  * [`2`:`scriptlen`]
+  * [`scriptlen`:`script`]
+
+1. subtype `output_info`
+2. data:
+  * [`8`:`satoshis`]
+  * [`2`:`scriptlen`]
+  * [`scriptlen`:`script`]
+
+#### Requirements
+
+The sending node:
+  - MUST ensure each `input_info` refers to an existing UTXO
+  - MUST ensure the `output_info`.`script` is a standard script
+  - MUST NOT spend the UTXOs it has specified (until/unless
+    channel establishment fails).
+  - MUST NOT include the channel funding output.
+  - if is the `opener`:
+    - MUST NOT send zero inputs (`num_inputs` cannot be zero).
+    - MUST specify an output with value zero, which will be used
+      as the change address.
+  - if is the `accepter`:
+    - consider the `put_limit` the total of `num_inputs` plus
+      `num_outputs` from `funding_compose`, with minimum 2.
+    - MUST NOT send a number of `input_info` or `output_info` which
+      exceeds the `put_limit`.
+    - MAY send zero inputs and/or outputs.
+
+The receiving node:
+  - if is the `opener`:
+    - MUST fail the channel if:
+      - the sum of `input_info` and `output_info` is greater than
+        the `put_limit`.
+  - otherwise:
+    - MUST fail the channel if:
+      - `change_satoshis` would be negative.
```

Comment:
It may not be able to figure this out yet (fees).  So, perhaps this should be 'sum of input amounts < sum of output amounts'?  And add the change_satoshis test only once it has calculated the tx?

Reply:
Hmm, actually, I think this being negative should be allowed: in that case, funding_satoshis should be reduced.  In effect, this allows you to say "I don't want change, take the fee out of my initial balance".  Or perhaps we should make the change output optional for this case?

Reply:
This seems like a valid use case. Will remove the '`change_satoshis` must be positive' requirement.

Allowing negative change amounts means the resulting channel balance will be less than the indicated `funding_satoshis`, which ostensibly came from the user. Implementations would probably want to add a way for users to indicate where to take the fees out -- either the provided channel fund amount or in addition to that. (You could do this with the existing flow too, since it's just UX.)

My hesitation with making the `change_address` optional is that 1) it's an edge case (depends on available UTXO set) 2) it has a small chance of introducing bugs (i.e. broken implementation forgets to include a change address so you end up with a huge channel balance by accident). Leaving out the change address doesn't save you much in calculation overhead when picking out the input UTXO set either. The only thing it does save on is wire bytes, in the case where your funding amt works out exactly right.